Error while processing screenshot testing result

### Expected behaviour
`debugExecuteScreenshotTests` should record screenshot tests, pull them from device and verify against the reference.

### Actual behaviour
After upgrading to the latest (5.12.2) version we started to receive the following exception while running screenshot tests:

```
> Task :library-module:connectedDebugAndroidTest
Starting 26 tests on Snapshotter(AVD) - 9

> Task :library-module:debugDownloadScreenshots
⬇️  Pulling screenshots from your connected devices!
Shot ADB warning: We could not pull screenshots from folder: /storage/emulated/0/Download/screenshots/com.app.test.uid/screenshots-default/
Shot ADB warning: rm: /storage/emulated/0/Download/screenshots/com.app.test.uid/screenshots-default/: No such file or directory

> Task :library-module:debugExecuteScreenshotTests FAILED
🔎  Comparing screenshots with previous ones.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':library-module:debugExecuteScreenshotTests'.
> com.fasterxml.jackson.databind.ObjectMapper.readerFor(Ljava/lang/Class;)Lcom/fasterxml/jackson/databind/ObjectReader;
```

We had to upgrade the version, because when we tried to migrate our screenshot test suite to use Compose, we started to receive the error mentioned in #244 , which was strange, because we are running this under Mac os, on emulator with SDK v28.

```
> Task :library-module:connectedDebugAndroidTest
Starting 26 tests on Snapshotter(AVD) - 9

> Task :library-module:debugDownloadScreenshots
⬇️  Pulling screenshots from your connected devices!
Shot ADB warning: We could not pull screenshots from folder: /storage/emulated/0/Download/screenshots/com.app.test.uid/screenshots-default/
Shot ADB warning: rm: /storage/emulated/0/Download/screenshots/com.app.test.uid/screenshots-default/: No such file or directory

> Task :library-module:debugExecuteScreenshotTests FAILED
🔎  Comparing screenshots with previous ones.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':library-module:debugExecuteScreenshotTests'.
> org.apache.commons.io.FileExistsException: Destination '/Users/user/StudioProjects/app/library-module/screenshots/debug/screenshots-default/com.app.test.uid.presentation.fragment.tool.elements.StepTest_testStepWithFallbackDescriptionPrimaryButton.png' already exists
```

### Steps to reproduce
- Create screenshot tests for Compose in library module
- try to run `debugExecuteScreenshotTests`
- observe error

### Version of the library
5.12.2
Mac OS 11.5.1
Java /Library/Java/JavaVirtualMachines/jdk-11.0.11.jdk/Contents/Home


**_BTW_**
Previously plugin saved screenshots in the folder `$module/screenshots/$flavor/` now after execution I see `screenshots-default` folder inside previously existing folder. And for my 55 tests were added 705 new files in this `screenshots-default` folder - one file `metadata.xml_emulator-5554`, 54 JSONs - like this `com.roadrunner.automationtool.presentation.fragment.automationtool.elements.ButtonTest_testDropdownButtonDisabled_dump.json`, 648 PNGs and one `screenshot_bundle.zip` with all this PNGs and JSONs. Is it ok?

--- a/core/src/main/scala/com/karumi/shot/Shot.scala
+++ b/core/src/main/scala/com/karumi/shot/Shot.scala
@@ -279,7 +279,7 @@ class Shot(
         )
       val screenshotSuite = metadataFiles.flatMap { metadataFilePath =>
         val metadataFileContent = files.read(metadataFilePath.getAbsolutePath)
-        ScreenshotsComposeSuiteJsonParser.parseScreenshots(
+        ScreenshotsComposeSuiteJsonParser.parseScreenshotSuite(
           metadataFileContent,
           shotFolder.screenshotsFolder(),
           shotFolder.pulledScreenshotsFolder(),

--- a/core/src/main/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParser.scala
+++ b/core/src/main/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParser.scala
@@ -1,5 +1,6 @@
 package com.karumi.shot.json
 
+import com.drew.metadata.Metadata
 import com.karumi.shot.domain.{Dimension, Screenshot}
 import com.karumi.shot.domain.model.{Folder, ScreenshotsSuite}
 import org.json4s._
@@ -7,15 +8,7 @@ import org.json4s.native.JsonMethods._
 
 object ScreenshotsComposeSuiteJsonParser {
 
-  def parseScreenshot(jsonNode: JsonAST.JValue): ComposeScreenshot = {
-    val JString(name)          = jsonNode \ "name"
-    val JString(testClassName) = jsonNode \ "testClassName"
-    val JString(testName)      = jsonNode \ "testName"
-
-    ComposeScreenshot(name, testClassName, testName)
-  }
-
-  def parseScreenshots(
+  def parseScreenshotSuite(
       metadataJson: String,
       screenshotsFolder: Folder,
       temporalScreenshotsFolder: Folder,
@@ -23,26 +16,64 @@ object ScreenshotsComposeSuiteJsonParser {
   ): ScreenshotsSuite = {
     implicit val formats: DefaultFormats.type = DefaultFormats
 
-    val json                            = parse(metadataJson)
-    val JObject(composeScreenshotSuite) = json
-    val JArray(composeScreenshots)      = json \ "screenshots"
+    val json = parse(metadataJson)
 
-    composeScreenshots.map(parseScreenshot).map { screenshot =>
-      val name = screenshot.name
-      Screenshot(
-        name = name,
-        recordedScreenshotPath = screenshotsFolder + name + ".png",
-        temporalScreenshotPath = screenshotsTemporalBuildPath + "/" + name + ".png",
-        testClass = screenshot.testClassName,
-        testName = screenshot.testName,
-        tilesDimension = Dimension(0, 0),
-        viewHierarchy = "",
-        absoluteFileNames = Seq(),
-        relativeFileNames = Seq(),
-        recordedPartsPaths = Seq(temporalScreenshotsFolder + "/" + name + ".png"),
-        screenshotDimension = Dimension(0, 0)
+    parseScreenshots(
+      json,
+      screenshotsFolder,
+      temporalScreenshotsFolder,
+      screenshotsTemporalBuildPath
+    )
+  }
+
+  private def parseScreenshots(
+      json: JValue,
+      screenshotsFolder: Folder,
+      temporalScreenshotsFolder: Folder,
+      screenshotsTemporalBuildPath: Folder
+  ): ScreenshotsSuite = {
+    val JArray(composeScreenshots) = json \ "screenshots"
+
+    composeScreenshots
+      .map(parseScreenshot)
+      .map(
+        mapComposeScreenshot(
+          _,
+          screenshotsFolder,
+          temporalScreenshotsFolder,
+          screenshotsTemporalBuildPath
+        )
       )
-    }
+  }
+
+  private def parseScreenshot(jsonNode: JsonAST.JValue): ComposeScreenshot = {
+    val JString(name)          = jsonNode \ "name"
+    val JString(testClassName) = jsonNode \ "testClassName"
+    val JString(testName)      = jsonNode \ "testName"
+
+    ComposeScreenshot(name, testClassName, testName)
+  }
+
+  private def mapComposeScreenshot(
+      screenshot: ComposeScreenshot,
+      screenshotsFolder: Folder,
+      temporalScreenshotsFolder: Folder,
+      screenshotsTemporalBuildPath: Folder
+  ): Screenshot = {
+    val name = screenshot.name
+    Screenshot(
+      name = name,
+      recordedScreenshotPath = screenshotsFolder + name + ".png",
+      temporalScreenshotPath = screenshotsTemporalBuildPath + "/" + name + ".png",
+      testClass = screenshot.testClassName,
+      testName = screenshot.testName,
+      tilesDimension = Dimension(0, 0),
+      viewHierarchy = "",
+      absoluteFileNames = Seq(),
+      relativeFileNames = Seq(),
+      recordedPartsPaths = Seq(temporalScreenshotsFolder + "/" + name + ".png"),
+      screenshotDimension = Dimension(0, 0)
+    )
   }
 }
 

--- a/core/src/main/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParser.scala
+++ b/core/src/main/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParser.scala
@@ -3,18 +3,31 @@ package com.karumi.shot.json
 import com.karumi.shot.domain.{Dimension, Screenshot}
 import com.karumi.shot.domain.model.{Folder, ScreenshotsSuite}
 import org.json4s._
-import org.json4s.jackson.JsonMethods._
+import org.json4s.native.JsonMethods._
 
 object ScreenshotsComposeSuiteJsonParser {
+
+  def parseScreenshot(jsonNode: JsonAST.JValue): ComposeScreenshot = {
+    val JString(name)          = jsonNode \ "name"
+    val JString(testClassName) = jsonNode \ "testClassName"
+    val JString(testName)      = jsonNode \ "testName"
+
+    ComposeScreenshot(name, testClassName, testName)
+  }
+
   def parseScreenshots(
-      json: String,
+      metadataJson: String,
       screenshotsFolder: Folder,
       temporalScreenshotsFolder: Folder,
       screenshotsTemporalBuildPath: Folder
   ): ScreenshotsSuite = {
     implicit val formats: DefaultFormats.type = DefaultFormats
-    val composeSuite                          = parse(json).extract[ComposeScreenshotSuite]
-    composeSuite.screenshots.map { screenshot =>
+
+    val json                            = parse(metadataJson)
+    val JObject(composeScreenshotSuite) = json
+    val JArray(composeScreenshots)      = json \ "screenshots"
+
+    composeScreenshots.map(parseScreenshot).map { screenshot =>
       val name = screenshot.name
       Screenshot(
         name = name,

--- a/core/src/test/resources/screenshots-metadata/compose-metadata.json
+++ b/core/src/test/resources/screenshots-metadata/compose-metadata.json
@@ -1,0 +1,19 @@
+{
+  "screenshots": [
+    {
+      "name": "com.example.test.ButtonTest_test1",
+      "testClassName": "com.example.test.ButtonTest",
+      "testName": "test1"
+    },
+    {
+      "name": "com.example.test.ButtonTest_test2",
+      "testClassName": "com.example.test.ButtonTest",
+      "testName": "test2"
+    },
+    {
+      "name": "com.example.test.ButtonTest_test3",
+      "testClassName": "com.example.test.ButtonTest",
+      "testName": "test3"
+    }
+  ]
+}

--- a/core/src/test/resources/screenshots-metadata/empty-compose-screenshots-metadata.json
+++ b/core/src/test/resources/screenshots-metadata/empty-compose-screenshots-metadata.json
@@ -1,0 +1,3 @@
+{
+  "screenshots": []
+}

--- a/core/src/test/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParserSpec.scala
+++ b/core/src/test/scala/com/karumi/shot/json/ScreenshotsComposeSuiteJsonParserSpec.scala
@@ -1,0 +1,103 @@
+package com.karumi.shot.json
+
+import com.karumi.shot.Resources
+import com.karumi.shot.domain.{Dimension, Screenshot}
+import com.karumi.shot.json.ScreenshotsComposeSuiteJsonParser.{
+  parseScreenshotSuite,
+  parseScreenshots
+}
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should
+
+import scala.Seq
+
+class ScreenshotsComposeSuiteJsonParserSpec
+    extends AnyFlatSpec
+    with should.Matchers
+    with Resources {
+  private val anyScreenshotsFolder = "/screenshots/"
+  private val anyTemporalScreenshotsFolder =
+    "/screenshots/screenshots-default/"
+  private val anyScreenshotsTemporalBuildPath =
+    "/build/tmp/shot/screenshots"
+
+  private val testClassName = "com.example.test.ButtonTest"
+  private val testName1     = "test1"
+  private val testName2     = "test2"
+  private val testName3     = "test3"
+  private val name1         = s"${testClassName}_${testName1}"
+  private val name2         = s"${testClassName}_${testName2}"
+  private val name3         = s"${testClassName}_${testName3}"
+
+  private val expectedScreenshots = Seq(
+    Screenshot(
+      name = name1,
+      recordedScreenshotPath = s"$anyScreenshotsFolder$name1.png",
+      temporalScreenshotPath = s"$anyScreenshotsTemporalBuildPath/$name1.png",
+      testClass = testClassName,
+      testName = testName1,
+      tilesDimension = Dimension(0, 0),
+      viewHierarchy = "",
+      absoluteFileNames = Seq(),
+      relativeFileNames = Seq(),
+      recordedPartsPaths = Seq(s"$anyTemporalScreenshotsFolder/$name1.png"),
+      screenshotDimension = Dimension(0, 0)
+    ),
+    Screenshot(
+      name = name2,
+      recordedScreenshotPath = s"$anyScreenshotsFolder$name2.png",
+      temporalScreenshotPath = s"$anyScreenshotsTemporalBuildPath/$name2.png",
+      testClass = testClassName,
+      testName = testName2,
+      tilesDimension = Dimension(0, 0),
+      viewHierarchy = "",
+      absoluteFileNames = Seq(),
+      relativeFileNames = Seq(),
+      recordedPartsPaths = Seq(s"$anyTemporalScreenshotsFolder/$name2.png"),
+      screenshotDimension = Dimension(0, 0)
+    ),
+    Screenshot(
+      name = name3,
+      recordedScreenshotPath = s"$anyScreenshotsFolder$name3.png",
+      temporalScreenshotPath = s"$anyScreenshotsTemporalBuildPath/$name3.png",
+      testClass = testClassName,
+      testName = testName3,
+      tilesDimension = Dimension(0, 0),
+      viewHierarchy = "",
+      absoluteFileNames = Seq(),
+      relativeFileNames = Seq(),
+      recordedPartsPaths = Seq(s"$anyTemporalScreenshotsFolder/$name3.png"),
+      screenshotDimension = Dimension(0, 0)
+    )
+  )
+
+  "ScreenshotsComposeSuiteJsonParser" should "return an empty spec if there are no screenshots" in {
+    val json = testResourceContent("/screenshots-metadata/empty-compose-screenshots-metadata.json")
+
+    val screenshots =
+      parseScreenshotSuite(
+        json,
+        anyScreenshotsFolder,
+        anyTemporalScreenshotsFolder,
+        anyScreenshotsTemporalBuildPath
+      )
+
+    screenshots shouldBe empty
+  }
+
+  it should "parse a regular metadata file" in {
+    val json = testResourceContent("/screenshots-metadata/compose-metadata.json")
+
+    val screenshots =
+      parseScreenshotSuite(
+        json,
+        anyScreenshotsFolder,
+        anyTemporalScreenshotsFolder,
+        anyScreenshotsTemporalBuildPath
+      )
+
+    screenshots.size shouldBe 3
+
+    screenshots should be(expectedScreenshots)
+  }
+}

